c/tskit/trees.c

```diff
@@ -2124,6 +2124,580 @@ tsk_treeseq_sample_count_stat(const tsk_treeseq_t *self, tsk_size_t num_sample_s
+
+                w_AB = tsk_bit_array_count(&ss_AB_samples);
+                w_A = tsk_bit_array_count(&ss_A_samples);
+                w_B = tsk_bit_array_count(&ss_B_samples);
+
+                hap_weight_row[0] = (double) w_AB;
+                hap_weight_row[1] = (double) (w_A - w_AB); // w_Ab
+                hap_weight_row[2] = (double) (w_B - w_AB); // w_aB
+            }
+            ret = f(state_dim, weights, result_dim, result_tmp_row, f_params);
+            if (ret != 0) {
+                goto out;
+            }
+            ret = norm_f(state_dim, weights, num_a_alleles - polarised_val,
+                num_b_alleles - polarised_val, norm, f_params);
+            if (ret != 0) {
+                goto out;
+            }
+            for (k = 0; k < state_dim; k++) {
+                result[k] += result_tmp_row[k] * norm[k];
+            }
+            result_tmp_row += state_dim; // Advance to the next column
+        }
+    }
+
+out:
+    tsk_safe_free(weights);
+    tsk_safe_free(norm);
+    tsk_safe_free(result_tmp);
+    tsk_bit_array_free(&ss_A_samples);
+    tsk_bit_array_free(&ss_B_samples);
+    tsk_bit_array_free(&ss_AB_samples);
+    tsk_bit_array_free(&AB_samples);
+    return ret;
+}
+
+static int
+get_all_mutation_samples(const tsk_treeseq_t *self, const tsk_size_t *site_offsets,
+    tsk_size_t *num_alleles, tsk_bit_array_t *allele_samples)
+{
+    int ret = 0;
+    tsk_tree_t tree;
+
+    ret = tsk_tree_init(&tree, self, TSK_TS_INIT_BUILD_INDEXES | TSK_NO_SAMPLE_COUNTS);
+    if (ret != 0) {
+        goto out;
+    }
+
+    for (ret = tsk_tree_first(&tree); ret == TSK_TREE_OK; ret = tsk_tree_next(&tree)) {
+        ret = get_mutation_samples(&tree, site_offsets, allele_samples, &num_alleles);
+        if (ret != 0) {
+            goto out;
+        }
+    }
+    // if adding code below, check ret before continuing
+
+out:
+    tsk_tree_free(&tree);
+    return ret;
+}
+
+static int
+tsk_treeseq_two_site_count_stat(const tsk_treeseq_t *self, tsk_size_t state_dim,
+    const tsk_bit_array_t *sample_sets, tsk_size_t result_dim, general_stat_func_t *f,
+    sample_count_stat_params_t *f_params, norm_func_t *norm_f,
+    const double *TSK_UNUSED(left_window), const double *TSK_UNUSED(right_window),
+    tsk_flags_t options, tsk_size_t *result_size, double **result)
+{
+    int ret = 0;
+
+    tsk_bit_array_t all_samples_bits, allele_samples, allele_samples_row;
+    tsk_bit_array_t site_a_state, site_b_state;
+    tsk_size_t site_id, num_alleles_cumsum, inner, result_offset;
+    tsk_size_t site_a, site_b;
+    bool polarised = false;
+
+    const tsk_size_t num_sites = self->tables->sites.num_rows;
+    const tsk_size_t num_samples = self->num_samples;
+    const tsk_size_t max_alleles = self->tables->mutations.num_rows + num_sites;
+    tsk_size_t *restrict site_offsets = tsk_malloc(num_sites * sizeof(*site_offsets));
+    tsk_size_t *restrict num_alleles = tsk_malloc(num_sites * sizeof(*num_alleles));
```

Comment:
There's no real point in making this a restrict pointer in this function, we're not doing any tight loops on it, and just passing it on to other functions. Declaring as a regular pointer means we can call ``tsk_safe_free`` and reduce the code a bit.

Reply:
Fixed.